### ISSUE TITLE
openssl cannot locate symbol SSLv2_client_method

### ISSUE THREAD

**Issue by robertvalik** [termux/termux-packages#12]
Hello,

I tried to configure mutt with smime support. Everything looks ok, but the actual encryption/decryption does nothing. This could be due to openssl returning some weird error:

$ openssl
CANNOT LINK EXECUTABLE: cannot locate symbol "SSLv2_client_method" referenced by "openssl"...

Any idea what could cause this?

**Comment by fornwall** [comment]
Hi, what is the output of `which openssl`?

It should be `/data/data/com.termux/files/usr/bin/openssl` which is installed by `apt install openssl-tool` - I'm suspecting the you pick up another openssl binary which is not compatible with the libssl.so provided by Termux..

**Comment by robertvalik** [comment]
Now this is interesting:

$ openssl
CANNOT LINK EXECUTABLE: cannot locate symbol "SSLv2_client_method" referenced by "openssl"...                                                                   $ which openssl
/data/data/com.termux/files/usr/bin/openssl
$ /data/data/com.termux/files/usr/bin/openssl
OpenSSL> quit

If I run the executable listed by which, it seems to be working. It seems the /system/bin/openssl is causing the problem... But that is in PATH later than the com.termux one.

**Comment by fornwall** [comment]
Did you find any more information on why `/system/bin/openssl` was picked up before the Termux on, despite it being later in `$PATH`?

**Comment by robertvalik** [comment]
No, I did not. And I cannot reproduce this behavior anymore...

**Comment by fornwall** [comment]
Ok, could it perhaps been a problem with bash caching the /system/bin/openssl in your path, even though you later installed the Termux one? See e.g. http://unix.stackexchange.com/questions/86012/what-is-the-purpose-of-the-hash-command

I'm closing the issue, reopen it if there is any problem

By the way, if you want to run binaries in /system/bin whose linked libraries clash with the Termux ones, you can reset the LD_LIBRARY_PATH as in:

    LD_LIBRARY_PATH= /system/bin/openssl

**Comment by rayluo** [comment]
FWIW, I was trying to install openssl on my termux environment today.

1. At the beginning, openssl was simply not there, and the prompt message suggests me to run literally `pkg install openssl-tool`.
2. So I did, and the installation seemed successful, but then when I ran openssl, it gave an error message similar to the one mentioned here in the OP's first post: `CANNOT LINK EXECUTABLE: cannot locate symbol "SSL_CTX_set_psk_server_callback" referenced by "openssl"`
3. With the hint of the current thread, I ran `apt install openssl`. And after the installation, I can run openssl now!

I'm documenting the process here, in case it would help others who encounter similar issue.

**Comment by Quasic** [comment]
That's strange. openssl is a dependency of openssl-tool, so it should have installed automatically. Perhaps there was an error during the first install...

**Comment by xeffyr** [comment]
Cannot reproduce this error on ARM and AArch64:
```
bash-4.4$ pkg install openssl-tool
Hit:1 http://termux.net stable InRelease
Reading package lists... Done
Building dependency tree       
Reading state information... Done
......
Unpacking openssl-tool (1.0.2n) ...
Setting up ca-certificates (20170920) ...
Setting up openssl (1.0.2n) ...
Setting up openssl-tool (1.0.2n) ...
bash-4.4$ openssl
OpenSSL>
```

**Comment by rayluo** [comment]
Thanks for your responses. I should probably mention that the issue I observed was on a Termux environment which I setup at maybe one year ago. And, because I disabled the Google Play Store auto update, the termux app on my Android device has never been updated since then. I did not feel a need to update, because my Termux environment works fine all the time anyway. The openssl problem might already be fixed since then.

So a follow-up question is probably "do I have to update my termux environment itself? If so, how to do that without starting from scratch"?

My current Termux environment:

    -bash-4.4$ uname -a
    Linux localhost 3.10.49-g3d12fcb #1 SMP PREEMPT Tue Jul 26 21:52:12 CST 2016 armv7l Android
